File: androidApp/app/src/test/java/com/ndefiorenze/samplemultiplatformkotlinapp/Object.kt
package com.ndefiorenze.samplemultiplatformkotlinapp

import com.ndefiorenze.KotlinObject
import org.junit.Assert
import org.junit.Test

class Object {

    @Test
    fun usageOfKotlinObject() {
        Assert.assertTrue(KotlinObject === KotlinObject)
    }

    @Test
    fun usageOfKotlinObject_properties() {
        playWithIntProperty()
        playWithNullableAnyProperty()
        playWithNullableStringProperty()
    }

    private fun playWithIntProperty() {
        Assert.assertTrue(KotlinObject.internalStateInt == KotlinObject.INITIAL_VALUE_FOR_INT_PROPERTY)
        KotlinObject.internalStateInt = 123
        Assert.assertTrue(KotlinObject.internalStateInt == 123)
    }

    private fun playWithNullableAnyProperty() {
        Assert.assertTrue(KotlinObject.internalStateNullableAny == KotlinObject.INITIAL_VALUE_FOR_NULLABLE_ANY_PROPERTY)
        KotlinObject.internalStateNullableAny = 123
        Assert.assertTrue(KotlinObject.internalStateInt == 123)
        KotlinObject.internalStateNullableAny = "Now I'm a string"
        Assert.assertTrue(KotlinObject.internalStateNullableAny == "Now I'm a string")
    }

    private fun playWithNullableStringProperty() {
        Assert.assertTrue(KotlinObject.internalStateNullableString == KotlinObject.INITIAL_VALUE_FOR_NULLABLE_STRING_PROPERTY)
        KotlinObject.internalStateNullableString = "Now I'm not null"
        Assert.assertTrue(KotlinObject.internalStateNullableString == "Now I'm not null")
    }

}

File: iosApp/iosAppTests/Object.swift
//
//  sdf.swift
//  iosAppTests
//
//  Created by Nicola De Fiorenze on 03/07/18.
//  Copyright © 2018 Nicola De Fiorenze. All rights reserved.
//

import XCTest
import KotlinLibrary

class Object: XCTestCase {
    
    func testKotlinObjectUsage() {
        XCTAssert(KotlinLibraryKotlinObject() === KotlinLibraryKotlinObject())
    }

    func testKotlinObjectUsage_properties() {
        playWithIntProperty()
        playWithNullableAnyProperty()
        playWithNullableStringProperty()
    }

    private func playWithIntProperty(){
        XCTAssert(KotlinLibraryKotlinObject().internalStateInt == KotlinLibraryKotlinObject().initial_VALUE_FOR_INT_PROPERTY)
        KotlinLibraryKotlinObject().internalStateInt = 123
        XCTAssert(KotlinLibraryKotlinObject().internalStateInt == 123)
    }

    private func playWithNullableAnyProperty(){
        XCTAssert((KotlinLibraryKotlinObject().internalStateNullableAny as! String) == KotlinLibraryKotlinObject().initial_VALUE_FOR_NULLABLE_ANY_PROPERTY )
        KotlinLibraryKotlinObject().internalStateNullableAny = 123
        XCTAssert((KotlinLibraryKotlinObject().internalStateNullableAny as! Int) == 123)
        KotlinLibraryKotlinObject().internalStateNullableAny = "Now I'm a string"
        XCTAssert((KotlinLibraryKotlinObject().internalStateNullableAny as! String) == "Now I'm a string")
    }

    private func playWithNullableStringProperty(){
        XCTAssert(KotlinLibraryKotlinObject().internalStateNullableString == (KotlinLibraryKotlinObject().initial_VALUE_FOR_NULLABLE_STRING_PROPERTY as! String?))
        KotlinLibraryKotlinObject().internalStateNullableString = "Now I'm not null"
        XCTAssert(KotlinLibraryKotlinObject().internalStateNullableString! == "Now I'm not null")
    }
    
}
